### ISSUE TITLE
Why model.forward() and model.classify() return different peaks? Which are expected to be more accurate? 

### ISSUE THREAD

**Issue by Krysia** [seisbench/seisbench#260]
Hello, 

We trained a custom model on our dataset using the peak-benchmark code and wanted to use the classify method from the seisbench model API. Is there any reason why this should not work? 

We observe different results when running model.forward method and model.model.classify method, where model is an instance of PhaseNetLit class. 

Also some difference can be observed when using the published PhaseNet model with the original weights, the peaks returned by the classify method are slightly different than peaks returned by the forward method on a short 30s sample. 

Below is the plot demonstrating the peaking results of a sample stream using our custom PhaseNet model  and original PhaseNet model. 

![image](https://github.com/seisbench/seisbench/assets/3986861/be274739-c827-4983-b485-53a4c7df1715)

The plot was produces with the code: 

```
def get_stations(stream): 
    stations = [tr.stats.station for tr in stream]
    return list(set(stations))
    
def get_max_peaks(picks): 
    max_p = - np.inf
    max_s = - np.inf
    pp, sp = None, None
    for p in picks: 
        if p.phase == 'P' and p.peak_value > max_p: 
            pp = p
            # max_p = pp.peak_value
        if p.phase == 'S' and p.peak_value > max_s:
            sp = p
            max_s = sp.peak_value

    return pp, sp
    
def normalize_stream(stream): 
    normalizer = sbg.Normalize(demean_axis=-1, amp_norm_axis=-1, amp_norm_type="peak")
    for j in range(3): 
            x = stream[j].data
            x = normalizer._demean(x)
            x = normalizer._detrend(x)
            x = normalizer._amp_norm(x)
            to_pad = 0
            if len(x)<3001: 
                to_pad = 3001 - len(x)
            x = np.pad(x, (0, to_pad), 'edge')
            stream[j].data = x[:3001]


def set_length(stream, length=3001): 
    for j in range(3): 
        x = stream[j].data
        to_pad = 0
        if len(x)<length: 
            to_pad = length - len(x)
        x = np.pad(x, (0, to_pad), 'edge')
        stream[j].data = x[:length]
         


def plot_predictions_for_stream(model, stream, stations): 
    
    fig = plt.figure(figsize=(15,len(stations)*4))
    ax = fig.subplots(len(stations)*3, 1, sharex=True)

    ### load model with orignal weigths for comparison
    model_ow = sbm.PhaseNet.from_pretrained("original")
    
    
    for i, station in enumerate(stations): 
        station_stream = stream.select(station=station)
        normalize_stream(station_stream)
        set_length(station_stream, 3001)

        for j, ch in enumerate(['E', 'N', 'Z']): 
            tr = station_stream[j]
            ax[i*3+j].plot(tr.times("matplotlib"), tr.data, "b-")
            ax[i*3+j].xaxis_date()
            ax[i*3+j].set_ylabel(f"{station} {tr.stats.channel}", rotation=0,  labelpad=20)
            fig.autofmt_xdate()

        ### find peaks with classify
        classify_outputs = model.model.classify(station_stream, P_threshold=0, S_threshold=0)
        pp, sp = get_max_peaks(classify_outputs)

        if pp is not None: 
            ax[i*3].axvline(pp.peak_time, lw=1, label='P (classify)', color='magenta')
        
        if sp is not None: 
            ax[i*3].axvline(sp.peak_time, lw=1, label='S (classify)', color='cyan')

        ### find peaks with classify using model with original weights
        classify_outputs_original_model = model_ow.classify(station_stream, P_threshold=0, S_threshold=0)
        ppo, spo = get_max_peaks(classify_outputs_original_model)

        if ppo is not None: 
            ax[i*3+1].axvline(ppo.peak_time, lw=1, label='P (classify) orignal weights', color='black')
        if spo is not None: 
            ax[i*3+1].axvline(spo.peak_time, lw=1, label='S (classify) original weights', color='grey')

        
        ### find peaks with forward
        tr_data = np.array([station_stream[0], station_stream[1], station_stream[2]])

        with torch.no_grad():
            model = model.float()
            pred = model.forward(torch.tensor(tr_data, device=model.device).unsqueeze(0).float())
            pred = pred[0].cpu().numpy()
            ppick_idx = np.argmax(pred[0])
            spick_idx = np.argmax(pred[1])

            ax[i*3].axvline(station_stream[0].times("matplotlib")[ppick_idx], lw=1, label='P (forward)', color='green')
            ax[i*3].axvline(station_stream[0].times("matplotlib")[spick_idx], lw=1, label='S (forward)', color='orange')

        ### find peaks with forward using model with original weights
        with torch.no_grad():
            model_ow = model_ow.float()
            pred = model_ow.forward(torch.tensor(tr_data, device=model_ow.device).unsqueeze(0).float())
            pred = pred[0].cpu().numpy()
            ppick_idx = np.argmax(pred[0])
            spick_idx = np.argmax(pred[1])

            ax[i*3+2].axvline(station_stream[0].times("matplotlib")[ppick_idx], lw=1, label='P (forward) original weights')
            ax[i*3+2].axvline(station_stream[0].times("matplotlib")[spick_idx], lw=1, label='S (forward) original weights')

        
        ax[i*3].legend(loc='center left')
        ax[i*3+1].legend(loc='center left')
        ax[i*3+2].legend(loc='center left')

stream = read('20210822045917.mseed')
model = models.PhaseNetLit().load_from_checkpoint("model.ckpt", map_location=torch.device('cpu') )
```

Could someone spot an issue or explain how to get the same output with the classify as with forward method? 
Could the different order of channels be the problem? (peak-benchmark code uses PSN order, while seisbench NSP)

Thank you very much, 
Krystyna

**Comment by yetinam** [comment]
Hi @Krysia ,
thank you for your report. I've gone through the code and there are a few things I'd recommend looking into:

- For PhaseNet original the normalisation uses std and not peak amplitude. This will change the outputs between forward (where no further normalisation is applied) and classify (where it is applied).
- The input component order for the original model is ENZ (as used) but for the other pretrained models it should be ZNE (see `model.component_order`). This seems to be incorrect in the forward call. 
- From the figure I guess it's the case, but you should double-check that the sampling rate in 100 Hz because classify will otherwise automatically resample (while forward will not).
- The model need to be set to evaluation mode before the forward calls (`model.eval()`). Otherwise the batch norm will behave differently. classify does this automatically.
- For the output channel order, the order can be accessed in `model.labels`.
- For the model loaded from a checkpoint, the hyperparameters might be wrong. You should check that `component_order`, `sampling_rate`, `labels` and `norm` are set appropriately (matching how the model was trained). These paramters are used internally for classify.
- The pick threshold at 0 is not advisable. This will lead to exactly one pick (per phase), starting at the beginning of the window and ending at the end of it. However, as you only access `peak_time`, this should still give you the time of the maximum.
- For debugging, I'd highly recommend you to plot the full probability curves output by annotate / of the forward function.
- As a last remark, the SeisBench models are generally not trained on padded data, so this might degrade performance. I'd always recommend using longer waveforms instead.

I hope these points help figuring out the source of the differences.

**Comment by zhong-yy** [comment]
Hi @Krysia and @yetinam, thank you for this informative discussion which is also beneficial to me. I didn't realize that the original phasenet uses the component order "ENZ" because it could work even when I assumed the order of "ZNE" despite a bit worse performance.

For @Krysia 's question about the channel order, I think the component order of the input and the channel order of the output are very important when you are using `forward` which takes an array as input. However, when you are using `classify` which takes an `obspy.Stream` object, the component order will be rearranged internally by the `annotate` function in the preprocessing stage.
 
Although the default output order is "NPS" for the PhaseNet class, this parameter can be overwritten by the `model_args` given in the _model.json.v1_ file when using `from_pretrained()` to load the model using the `from _pretrained` function which calls `load()` https://github.com/seisbench/seisbench/blob/61ac2df822249fc16b62dfe7da1779ec2e8255bf/seisbench/models/base.py#L536-L537

I suppose component order of an input stream is re-arranged by `annotate` somewhere, but I haven't found where this ordering happens in the code.

**Comment by yetinam** [comment]
Thanks @zhong-yy for the great additions. For the rearranging of input components in `annotate`, that's happening here: https://github.com/seisbench/seisbench/blob/61ac2df822249fc16b62dfe7da1779ec2e8255bf/seisbench/models/base.py#L2371-L2373

**Comment by zhong-yy** [comment]
Thank you for your kind quotation @yetinam. 

> Thanks @zhong-yy for the great additions. For the rearranging of input components in `annotate`, that's happening here:
> 
> https://github.com/seisbench/seisbench/blob/61ac2df822249fc16b62dfe7da1779ec2e8255bf/seisbench/models/base.py#L2371-L2373

**Comment by Krysia** [comment]
Hi @yetinam and @zhong-yy, 

Thank you @yetinam for taking time to go through the code and suggesting solutions. Thank you @zhong-yy for joining the discussion. 

Unifying component order, normalization type ('peak' vs 'std'), label order solved the problem for seisbench PhaseNet model with original weights. The output of forward and classifiy is nearly identical. I am still struggling with achieving the same with the custom model (a pytorch lightning object), trained with the pick-benchmark code. Forward called directly on the model (PL object) still produces different results than classify called from model.model (the native seisbench Phasenet object). Sometimes it is very close, sometimes it's completely different. 

The output of forward is closer to what is expected, because it uses settings applied during training. I guess there are some other parameters, which have different default values in pick-benchmark and original seisbench repo. Maybe overlap? Is there anything else I should examine? 

Thank you again!

**Comment by yetinam** [comment]
Hi @Krysia ,
thanks for the update. Glad to see that it work now at least with the PhaseNet model with original weights. I'm not sure where the differences could come from if you addressed all points from my earlier post. Overlap in general should only be an issue if your trace is longer than the input window, i.e., will be chunked. If your trace is exactly 3001 samples at 100 Hz that should not happen. Can you make out any patterns, when the differences happen? I'd personally suspect there is something different in the preprocessing, like a different epsilon value in normalisation. Do your input waveforms maybe have very small amplitudes?

**Comment by Krysia** [comment]
Hi @yetinam, 

Finally all works fine. There were several issues, I will describe them in case someone struggles with something similar. 
* My input is longer than a window length, I was performing normalisation before slicing when using forward method, whereas classify method normalises each window separately, so no wonders that the effect was different, so normalization first, next slicing
* In my dataset some streams had order ENZ, others ZNE, sorting traces solves this issue
* Some traces had very slightly different starttime, so syncing them was needed. 

Thank you for help.

**Comment by yetinam** [comment]
Hi @Krysia ,
great to hear that the issue is resolved and thanks a lot for detailing the solution here!
